File: src/main/java/AoC2015_03.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;
import static java.util.stream.Collectors.toList;

import java.util.List;
import java.util.stream.Stream;

import com.github.pareronia.aoc.Utils;
import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aoc.navigation.Heading;
import com.github.pareronia.aoc.navigation.NavigationWithHeading;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2015_03 extends SolutionBase<List<Direction>, Long, Long> {

    private AoC2015_03(final boolean debug) {
        super(debug);
    }

    public static AoC2015_03 create() {
        return new AoC2015_03(false);
    }

    public static AoC2015_03 createDebug() {
        return new AoC2015_03(true);
    }
 
    @Override
    protected List<Direction> parseInput(final List<String> inputs) {
        assert inputs.size() == 1;
        return Utils.asCharacterStream(inputs.get(0))
                .map(Direction::fromChar).collect(toList());
    }

    @Override
    public Long solvePart1(final List<Direction> input) {
        final HouseVisits houseVisits = new HouseVisits();
        input.forEach(houseVisits::goVisit);
        return houseVisits.getUniqueVisits().count();
    }

    @Override
    public Long solvePart2(final List<Direction> input) {
        final HouseVisits santaVisits = new HouseVisits();
        range(0, input.size(), 2).intStream()
                .mapToObj(input::get)
                .forEach(santaVisits::goVisit);
        final HouseVisits robotVisits = new HouseVisits();
        range(1, input.size(), 2).intStream()
                .mapToObj(input::get)
                .forEach(robotVisits::goVisit);
        return Stream.concat(
                    santaVisits.getUniqueVisits(), robotVisits.getUniqueVisits())
                .distinct().count();
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "2"),
        @Sample(method = "part1", input = TEST2, expected = "4"),
        @Sample(method = "part1", input = TEST3, expected = "2"),
        @Sample(method = "part2", input = TEST4, expected = "3"),
        @Sample(method = "part2", input = TEST2, expected = "3"),
        @Sample(method = "part2", input = TEST3, expected = "11"),
    })
    public void samples() {
    }

    public static void main(final String[] args) throws Exception {
        AoC2015_03.create().run();
    }

    private static final String TEST1 = ">";
    private static final String TEST2 = "^>v<";
    private static final String TEST3 = "^v^v^v^v^v";
    private static final String TEST4 = "^v";
    
    private static final class HouseVisits {
        private final NavigationWithHeading nav
                = new NavigationWithHeading(Position.ORIGIN, Heading.NORTH);
        
        public void goVisit(final Direction direction) {
            this.nav.navigate(Heading.fromDirection(direction), 1);
        }
        
        public Stream<Position> getUniqueVisits() {
            return nav.getVisitedPositions(true).stream().distinct();
        }
    }
}

File: src/main/cpp/2015/03/AoC2015_03.cpp
#include <assert.h>

#include <string>
#include <unordered_set>
#include <vector>

#include "../../aoc/aoc.hpp"
#include "../../aoc/grid/grid.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;

Cell move(const Cell& curr, const char& ch) {
    if (ch == '^') {
        return Cell::at(curr.row() - 1, curr.col()); 
    } else if (ch == '>') {
        return Cell::at(curr.row(), curr.col() + 1); 
    } else if (ch == 'v') {
        return Cell::at(curr.row() + 1, curr.col()); 
    } else if (ch == '<') {
        return Cell::at(curr.row(), curr.col() - 1); 
    } else {
        throw "Invalid input";
    }
}

int part1(const vector<string>& input) {
    Cell curr = Cell::at(0, 0);
    unordered_set<Cell> visited({curr});
    for (const char& ch : input [0]) {
        curr = move(curr, ch);
        visited.insert(curr);
    }
    return visited.size();
}

int part2(const vector<string>& input) {
    Cell currS = Cell::at(0, 0);
    Cell currR = Cell::at(0, 0);
    unordered_set<Cell> visited({currS, currR});
    for (const int i : aoc::Range::range(input[0].size())) {
        const char& ch = input[0][i];
        if (i % 2== 0) {
            currS = move(currS, ch);
            visited.insert(currS);
        } else {
            currR = move(currR, ch);
            visited.insert(currR);
        }
    }
    return visited.size();
}

const vector<string> TEST1 = {">"};
const vector<string> TEST2 = {"^>v<"};
const vector<string> TEST3 = {"^v^v^v^v^v"};
const vector<string> TEST4 = {"^v"};

void samples() {
    assert(part1(TEST1) == 2);
    assert(part1(TEST2) == 4);
    assert(part1(TEST3) == 2);
    assert(part2(TEST4) == 3);
    assert(part2(TEST2) == 3);
    assert(part2(TEST3) == 11);
}

MAIN(2015, 3)
